### PR TITLE
[ONNX] Update the default opset to version 14

### LINKED ISSUE #81075: [ONNX] Support opset 17 operators
Tracking issue on the onnx opset 17 support. Contributions welcome.

- [x] Add version 17 as a known version #83287 
- [ ] fft
  - [ ] ifft
  - [ ] fft_fft
  - [ ] fft_ifft
  - [ ] fft_rfft
  - [ ] fft_irfft
  - [ ] fft_hfft
  - [ ] fft_ihfft
  - [ ] fft_fft2
  - [ ] fft_ifft2
  - [ ] fft_rfft2
  - [ ] fft_irfft2
  - [ ] fft_hfft2
  - [ ] fft_ihfft2
  - [ ] fft_fftn
  - [ ] fft_ifftn
  - [ ] fft_rfftn
  - [ ] fft_irfftn
  - [ ] fft_hfftn
  - [ ] fft_ihfftn
  - [ ] fft_fftfreq
  - [ ] fft_rfftfreq
  - [ ] fft_fftshift
  - [ ] fft_ifftshift
- [x] stft #83944 
- [ ] istft
- [ ] bartlett_window
- [ ] blackman_window
- [ ] hamming_window
- [ ] hann_window
- [ ] kaiser_window
- [x] LayerNormalization #84293

Closes #80834

### PR DESCRIPTION
**Pull request by justinchuby**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #83284
* #83283

Update the default opset by running the `update_default_opset_version.py` script. The update is done in a regularly to ensure we are in sync with the onnx updates. All changes are produced by the script.

### CODE REVIEW

**Review by thiagocrepaldi — COMMENTED** [review]
Hi @justinchuby, please provide a clear and concise description of why we are changing the default opset. Documenting the reason helps us in the future, when we are debugging problems that are specific to opset version that affects a particular runtime

What is the motivation for the proposal? Changing the default opset can break Caffe2 and any other runtime that consumes. IIRC, it might be because we are trying to stay 3 version behind the latest, is that it?

File: test/onnx/expect/TestOperators.test_view.expect [hunk 1]
```diff
@@ -22,6 +22,11 @@ graph {
     output: "2"
     name: "Reshape_1"
     op_type: "Reshape"
+    attribute {
+      name: "allowzero"
+      i: 0
+      type: INT
+    }
```
Comment (thiagocrepaldi) [inline test/onnx/expect/TestOperators.test_view.expect:29]: are these expected with opset 14?

**Comment by justinchuby** [comment]
> Hi @justinchuby, please provide a clear and concise description of why we are changing the default opset. Documenting the reason helps us in the future, when we are debugging problems that are specific to opset version that affects a particular runtime
> 
> What is the motivation for the proposal? Changing the default opset can break Caffe2 and any other runtime that consumes. IIRC, it might be because we are trying to stay 3 version behind the latest, is that it?

Done. PTAL.

The change is done by running the script. As I understand it - yes it is because we want to use the oldest version in the last 18 months (mentioned in the script).

If versions are usually explicitly specified during conversion, I am hoping nothing will break with this change?

File: test/onnx/expect/TestOperators.test_view.expect (hunk 1 above)
Comment (justinchuby) [reply test/onnx/expect/TestOperators.test_view.expect:29]: I think so? They are automatically generated.

Comment (justinchuby) [reply test/onnx/expect/TestOperators.test_view.expect:29]: I found the documented change in https://github.com/onnx/onnx/blob/main/docs/Changelog.md#Reshape-14

**Comment by thiagocrepaldi** [comment]
> > Hi @justinchuby, please provide a clear and concise description of why we are changing the default opset. Documenting the reason helps us in the future, when we are debugging problems that are specific to opset version that affects a particular runtime
> > What is the motivation for the proposal? Changing the default opset can break Caffe2 and any other runtime that consumes. IIRC, it might be because we are trying to stay 3 version behind the latest, is that it?
> 
> Done. PTAL.
> 
> The change is done by running the script. As I understand it - yes it is because we want to use the oldest version in the last 18 months (mentioned in the script).
> 
> If versions are usually explicitly specified during conversion, I am hoping nothing will break with this change? Do we need an import to verify?

Assuming people use explicit version on the torch.onnx.export call, nothing should break, but the API also allows not setting it, so a break is possible. Although the ONNX converter team know about the policy for updating the opset, the general public probably dont, so expliciting the reason on the PR is always a good idea. Especially for those that do not explicitly specify opset version and spend days debugging a model change they have no idea where it came from :P

**Review by thiagocrepaldi — COMMENTED** [review]
The change LGTM.

One aspect that we never discussed when changing opset versions before, but I thought about it now is: Should we add `module: BC-breaking` or `topic: BC breaking` Github labels on the PR that changes opset version? 

For scripts that use implicit opset version, the opset change can cause BC issues and the PyTorch release notes should explicit that

@BowenBao @shubhambhokare1 any thoughts?

**Comment by BowenBao** [comment]
> The change LGTM.
> 
> One aspect that we never discussed when changing opset versions before, but I thought about it now is: Should we add `module: BC-breaking` or `topic: BC breaking` Github labels on the PR that changes opset version?
> 
> For scripts that use implicit opset version, the opset change can cause BC issues and the PyTorch release notes should explicit that
> 
> @BowenBao @shubhambhokare1 any thoughts?

I agree with the BC breaking concern and I think it's the cost we have to pay for moving forward. Good idea to add the BC-breaking tags. Linking the original PR from Gary for reference #73898.

@justinchuby I'm not sure about this line `Update the default opset ... to prepare for opset 17 support`. Why are these related? If there is no dependency I'd actually recommend separating this PR, as it might get reverted for failing Meta internal tests.

**Comment by justinchuby** [comment]
> I agree with the BC breaking concern and I think it's the cost we have to pay for moving forward. Good idea to add the BC-breaking tags. Linking the original PR from Gary for reference #73898.
> 
> @justinchuby I'm not sure about this line `Update the default opset ... to prepare for opset 17 support`. Why are these related? If there is no dependency I'd actually recommend separating this PR, as it might get reverted for failing Meta internal tests.

Thanks for pointing this out. For some reason I just assumed it is the process when there is a new opset coming out. I have detached #83287

**Review by titaiwangms — APPROVED** [review]
🍺

File: test/onnx/expect/TestOperators.test_view.expect (hunk 1 above)
Comment (justinchuby) [reply test/onnx/expect/TestOperators.test_view.expect:29]: Marking as resolved

**Review by malfet — APPROVED** [review]
LGTM, not sure why it has `onnx-needs-import` label

**Review by BowenBao — APPROVED** [review]

**Pull request closed without merging**